File: leetcode/leetcode-56.py
import functools
from typing import List

class Solution:
    def merge(self, intervals: List[List[int]]) -> List[List[int]]:
        if len(intervals) <= 0:
            return []
        def cmp(a, b):
            if a[0] == b[0]:
                return -1 if a[1] <= b[1] else 1
            return -1 if a[0] <= b[0] else 1
        intervals = sorted(intervals, key = functools.cmp_to_key(cmp))
        ans = []
        l = intervals[0][0]
        r = intervals[0][1]
        for i in range(1, len(intervals)):
            if intervals[i][0] > r:
                ans.append([l, r])
                l = intervals[i][0]
                r = intervals[i][1]
            r = max(r, intervals[i][1])
        ans.append([l, r])
        return ans

if __name__ == "__main__":
    so = Solution()
    print(so.merge([[1,4],[0,4]]))

File: leetcode/leetcode-56.cpp
/*
 *@File Name          :leetcode-56.cpp
 *@Author             :cc
 *@Version            :1.0.0
 *@Date               :2022/3/5 9:00
 *@Description        :
 *@Function List      :
 *@History            :
 */

#include "vector"
#include "algorithm"

using namespace std;

class Solution {
public:
    vector<vector<int>> merge(vector<vector<int>> &intervals) {
        vector<vector<int>> ans;
        sort(intervals.begin(), intervals.end(), [](const vector<int> &a, const vector<int> &b) -> bool {
            if (a[0] == b[0]) return a[1] > b[1];
            return a[0] < b[0];
        });
        int pre = -1;
        int st;
        for (auto &x: intervals) {
            if (x[0] > pre) {
                if (pre != -1)
                    ans.push_back({st, pre});
                st = x[0];
                pre = x[1];
            } else {
                pre = max(pre, x[1]);
            }
        }
        ans.push_back({st, pre});
        return ans;
    }
};